chips/rp2040/src/pio.rs

```diff
@@ -0,0 +1,1231 @@
+const PIO1_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new(PIO_1_BASE_ADDRESS as *const PioRegisters) };
+const PIO1_XOR_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x1000) as *const PioRegisters) };
+const PIO1_SET_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x2000) as *const PioRegisters) };
+const PIO1_CLEAR_BASE: StaticRef<PioRegisters> =
+    unsafe { StaticRef::new((PIO_1_BASE_ADDRESS + 0x3000) as *const PioRegisters) };
+
+// There are a total of 4 State Machines per PIO.
+#[derive(Clone, Copy, PartialEq, Debug)]
+pub enum SMNumber {
+    SM0 = 0,
+    SM1 = 1,
+    SM2 = 2,
+    SM3 = 3,
+}
+
+// There can be 2 PIOs per RP2040.
+#[derive(PartialEq)]
+pub enum PIONumber {
+    PIO0 = 0,
+    PIO1 = 1,
+}
+
+impl RPGpio {
+    fn from_u32(value: u32) -> RPGpio {
+        match value {
+            0 => RPGpio::GPIO0,
+            1 => RPGpio::GPIO1,
+            2 => RPGpio::GPIO2,
+            3 => RPGpio::GPIO3,
+            4 => RPGpio::GPIO4,
+            5 => RPGpio::GPIO5,
+            6 => RPGpio::GPIO6,
+            7 => RPGpio::GPIO7,
+            8 => RPGpio::GPIO8,
+            9 => RPGpio::GPIO9,
+            10 => RPGpio::GPIO10,
+            11 => RPGpio::GPIO11,
+            12 => RPGpio::GPIO12,
+            13 => RPGpio::GPIO13,
+            14 => RPGpio::GPIO14,
+            15 => RPGpio::GPIO15,
+            16 => RPGpio::GPIO16,
+            17 => RPGpio::GPIO17,
+            18 => RPGpio::GPIO18,
+            19 => RPGpio::GPIO19,
+            20 => RPGpio::GPIO20,
+            21 => RPGpio::GPIO21,
+            22 => RPGpio::GPIO22,
+            23 => RPGpio::GPIO23,
+            24 => RPGpio::GPIO24,
+            25 => RPGpio::GPIO25,
+            26 => RPGpio::GPIO26,
+            27 => RPGpio::GPIO27,
+            28 => RPGpio::GPIO28,
+            29 => RPGpio::GPIO29,
+            _ => panic!(
+                "Unknown value for GPIO pin: {} (should be from 0 to 29)",
+                value
+            ),
+        }
+    }
+}
+
+// The FIFO queues can be joined together for twice the length in one direction.
+#[derive(PartialEq)]
+pub enum PioFifoJoin {
+    PioFifoJoinNone = 0,
+    PioFifoJoinTx = 1,
+    PioFifoJoinRx = 2,
+}
+
+const STATE_MACHINE_NUMBERS: [SMNumber; NUMBER_STATE_MACHINES] =
+    [SMNumber::SM0, SMNumber::SM1, SMNumber::SM2, SMNumber::SM3];
+
+pub struct Pio {
+    registers: StaticRef<PioRegisters>,
+    pio_number: PIONumber,
+    xor_registers: StaticRef<PioRegisters>,
+    set_registers: StaticRef<PioRegisters>,
+    _clear_registers: StaticRef<PioRegisters>,
+}
+
+// 'MOV STATUS' types.
+#[derive(Clone, Copy)]
+pub enum PioMovStatusType {
+    StatusTxLessthan = 0,
+    StatusRxLessthan = 1,
+}
+
+// PIO State Machine configuration structure
+//
+// Used to initialize a PIO with all of its state machines.
+pub struct StateMachineConfiguration {
```

Comment:
All of these struct comments should be `///`.

Reply:
Thank you!